### DISCUSSION TITLE
Question about EMF Corrections and CT

### DISCUSSION THREAD

**Discussion by pdmullen** [PrincetonUniversity/athena#455] [General]
Hi, everyone! I have been thinking about EMF corrections recently and wanted to run a question by everyone:

For uniform meshes (without any refinement), Athena++ communicates EMFs between neighboring MeshBlocks and averages them.  This is because when computing the EMF at boundaries for two different neighboring MeshBlocks, round-off error level differences can arise.  My question, however, is…

After communicating and averaging EMF’s at MeshBlock boundaries, those updated EMF’s are used to perform CT.  _Can CT also produce round-off error level differences in the updated magnetic field at a face that is shared by two MeshBlock boundaries?_ If so, I worry that in highly chaotic simulations, the magnetic field value at the same face but from the perspective of two different MeshBlocks may begin to differ.  Is this possible?

If this was an issue (which I am not convinced it is), the solution would be to just communicate the magnetic field component at the face shared by two MeshBlocks and average them (just like the EMFs).  The resultant averaged magnetic field should still be divergence-less to round-off error.

**Comment by zhuzh1983** [comment]
I think the problem is that without doing so, EMF at the edges will never communicate between two different cores and their difference can grow in time. Eventually, the difference grows exponentially and is so large that the code crashed. Theoretically we don't have to do it at any step. For B and CT, if we only have a little bit round-off error at one step, this may not be an issue. If we just average B without correcting EMF, the EMF error can grow exponentially.

**Reply by pdmullen** [reply]
Thanks, @zhuzh1983! I guess I am proposing that one might require **both** the EMF correction and the post-CT B field component averaging at the shared face.

**Reply by pdmullen** [reply]
To take a very, very extreme example to illustrate the point:  

I worry that grossly unfortunate growth of round-off error could lead to one `MeshBlock`'s cell that shares a face (let's say at `x3f = 0`) with another `MeshBlock` cell may say the z-component of the field is 1 kG, whereas from the perspective of the neighbor `MeshBlock`, the field strength at the same physical location is 1 G.  This could lead to some very weird magnetic tension forces, particularly in high sigma flows.

**Reply by zhuzh1983** [reply]
Do we communicate face B value of the active zone interface after each cycle? I thought the code did so.  If that is the case, within one step before the communicate, the round-off error between two B will be at the level of 1.e-15. Or you are worrying about the ghost zone B value?

**Reply by pdmullen** [reply]
I thought we did not communicate the B value of the active zone interface shared by two MeshBlocks at the same refinement level. If I am wrong, then this answers my question :-)

**Reply by zhuzh1983** [reply]
I am not sure either... Somebody please :)

**Reply by tomidakn** [reply]
No, we do not send/receive the active surface B-field that is shared by two MeshBlocks. This is because manipulating the active B-field causes violation of divB=0.

**Reply by zhuzh1983** [reply]
Then, following the same logic, we shouldn't copy and average B1?

**Reply by tomidakn** [reply]
Well, as far as the error remains as small as the machine epsilon, it should be fine. But that's why I implemented only EMF correction and did not touch the face field itself.

**Reply by pdmullen** [reply]
> No, we do not send/receive the active surface B-field that is shared by two MeshBlocks. This is because manipulating the active B-field causes violation of divB=0.

The post-CT B-fields at neighboring same-level `MeshBlock`s shared faces should be equal to ~round-off error.  Averaging them should still keep divB to zero to machine precision, right?

**Reply by tomidakn** [reply]
I guess so, but if the change by the averaging is larger than (or on the same order of) the error in the CT, it might cause a new problem.

Since I have never experienced this issue by myself, I am not sure if averaging really helps. I think we only can tell by experiments with a reproducible test case.

**Comment by tomidakn** [comment]
I guess it is theoretically possible that the error in the CT phase may cause mismatch in the shared face field, but I have never seen such a case in my experience, including turbulent protoplanetary disks in spherical polar coordinates.
